### ISSUE TITLE
Cookies lose implicit port after having port property set to empty string.

### ISSUE THREAD

**Issue by dsisco11** [dotnet/runtime#70227]
### Description

Using System.Text.Json to serialize and deserialize System.Net.Cookie objects will not always maintain roundtrip integrity.   
   
Specifically, cookies with an implicit port will lose their 'm_port_implicit' flag after deserialization!   
When a cookie is being serialized, any cookie with an implicit port will have the value string.Empty for the 'Port' property.   
So the Json serializer will naturally serialize that empty string value.   
The problem is that upon deserialization the deserializer will then set the Port property on the cookie to an empty string.   
Setting the 'Port' property of a cookie to ANY value currently removes the implicit port flag!   
   
The result of this is that SOME deserialized cookies will not match the Http requests that they should.   
   
Previously it was common to use a binary serializer for storing cookie containers, however the binary serializer was obsoleted due to security concerns.   
   
Seeing as how Json has become a popular standard for handling data serialization in general, I think this issue is actually fairly concerning as this behavior makes it impractical to handle cookies with the Json serializer.

### Reproduction Steps

```c#
var cookieA = new System.Net.Cookie("session", "foobar", "/", "foo.bar");
var cookieJson = System.Text.Json.JsonSerializer.Serialize(cookieA);
var cookieB = System.Text.Json.JsonSerializer.Deserialize<System.Net.Cookie>(cookieJson);

// Add cookie A to a container
var cookiesA = new System.Net.CookieContainer();
cookiesA.Add(cookieA);

// Add cookieB to a container
var cookiesB = new System.Net.CookieContainer();
cookiesB.Add(cookieB);

// Test uri matching
Uri testUri = new Uri("https://foo.bar/account/login");

var headA = cookiesA.GetCookieHeader(testUri);
var headB = cookiesB.GetCookieHeader(testUri);

Assert.Equal(headA, headB);
```

### Expected behavior

The returned cookie header string for A&B should match, because A&B should both match the provided uri.

### Actual behavior

The header string for the deserialized cookie (B) is empty because B no longer matches to the uri.

### Other information

I would suggest that the 'Port' property setter of the System.Net.Cookie class be updated such that 'm_port_implicit' is set to true when the value provided to the setter is NULL, but keep the current behavior when the value is an empty string.   
   
In the event that a cookies Port property is assigned a null or empty value it will not be matched to any requests as it has neither a valid list of port values nor the implicit port flag (See: 'MergeUpdateCollections' private function within System.Net.CookieContainer).    
So the current behavior of setting 'm_implicit_port' to false when provided a null or empty string doesn't make sense as it just kills the cookie.

**Comment by huoyaoyuan** [comment]
It seems that `Cookie` in fact does not round-trip its state upon property-by-property copying. Any serializer that uses property would be affected.

**Comment by dsisco11** [comment]
I have been attempting to come up with alternative solutions to fix this.   
I toyed with the idea of a new property that allowed explicitly specifying the 'Port' property as being implicit, but that solution is rendered ineffective because it would be overwritten in the event that the 'Port' property value were deserialized after the new property.   
Since the order of property serialization cannot be guaranteed, this cannot be relied on.   
   
I know that this is a well established component that already has a lot of other code which uses it and which may be affected by the change.   
But this is a fundamental issue and it is fundamentally broken in this manner.   
   
I'm not entirely clear about all of the RFC cookie specifications but do the standards dictate that a 'set-cookie' header which includes an empty value for 'Port' results in a cookie whose port is NOT implicit?   
It only makes sense that any cookie which does not have an ACTUAL set of port values supplied would then have an implicit port, does it not?   
Because in the event that a 'set-cookie' header specifies Port="" then the resultant cookie will have neither an implicit port nor valid port values, rendering it unmatchable for any domain or request.   
So shouldn't a null or empty string therefore result in an implicit port?   
   
   
I can only think of one other work around to allow for serialization without using the obsoleted binary serializers.   
Which is to provide two new functions, one to get an existing cookies equivalent 'set-cookie' header string, and another static extension function to directly instantiate a cookie from a 'set-cookie' header string.   
This would allow for cookies to be serialized/deserialized and stored as a series of 'set-cookie' strings.   
   
Again though, this isn't a valid solution to the issue, merely a temporary work around. The fundamental issue should still be addressed.   
   
As a side note I'm not sure why System.Net.Cookie doesn't already have a function for translating it into a valid 'set-cookie' header string and a function to directly instantiate a new cookie from a 'set-cookie' header string.

**Comment by antonfirsov** [comment]
@krwq @eiriktsarpalis can users expect that these kind of old/heavy/complicated BCL objects roundtrip with System.Text.Json? In other words: is this issue a bug, or is serialization of BCL objects something we discourage?

**Comment by eiriktsarpalis** [comment]
If the type was designed to be roundtripped using private field serialization, then yes, that is something System.Text.Json will not support. Unless the type can be changed somewhat to make roundtripping using properties feasible, our general recommendation in such cases is to author a custom converter that provides the right serialization semantics for the type.

**Comment by karelz** [comment]
Triage: We are not sure if it is good idea to work around the problem in Cookie setter.
Does not seem as high pri for 7.0 (only 1 report). Moving to 8.0 for now.

**Comment by dsisco11** [comment]
> If the type was designed to be roundtripped using private field serialization, then yes, that is something System.Text.Json will not support. Unless the type can be changed somewhat to make roundtripping using properties feasible, our general recommendation in such cases is to author a custom converter that provides the right serialization semantics for the type.

The System.Net.Cookie type CAN be changed to make roundtripping via properties possible.   
The problem with the type is that setting an empty value for Cookie.Port causes the cookie to flag the port as non-implicit.   

But my retort here is that this behavior in and of itself makes absolutely no sense. An empty port value within a cookie currently makes the cookie unusable as it can no longer be matched to ANY URI.   
The reason for this is that in order to match a cookie to a URI, the port of the URI must match the cookie.   
Normally a cookie with no explicit numeric values given for its port will implicitly assume the cookies port to be 80 (this is the correct behavior).   
However; The System.Net.Cookie type does not implicitly assume the port if it has internally flagged itself as non-implicit.   
   
So you see the problem is that any attempt to specify no value or an empty value for Cookie.Port counterintuitively causes it to both have no port values and also be incapable of assuming the port value, which it SHOULD be doing if no port value was given.   
   
Therefore, System.Net.Cookie SHOULD mark its port as implicit if it is given an empty or null string for the Port property.

**Comment by antonfirsov** [comment]
@dsisco11 makes sense to me, should be easy to fix by moving lines around in the setter code:

https://github.com/dotnet/runtime/blob/9b4f152802b6305438bc1f6ddf4733104bb57f6e/src/libraries/System.Net.Primitives/src/System/Net/Cookie.cs#L589-L597

Any chance you are interested raising a PR?

**Comment by dsisco11** [comment]
Yea I'll do a PR!
Sorry for the late response, Been really busy because I just got a new job and moved across the USA lol.

**Comment by dsisco11** [comment]
I filed a PR.
https://github.dev/dotnet/runtime/pull/76143

**Comment by dsisco11** [comment]
I updated the title to better suit the issue.
I'd like to note that I strongly disagree with with being considered a non-bug or enhancement.
That's partially my own fault for having not explained the true depth of this issue at the start.

To clearly state things; Cookie strings which specify an empty port will result in a corrupted cookie.
I say corrupted because the resulting instance of System.Net.Cookie will be non-functional and incapable of matching with ANY uri or http request.

Specifying an empty string for a cookies port is perfectly legal and there are many instances where a webserver will provide such cookie strings.
The correct behavior is for the cookie to assume the port implicitly to be the default port (80).
All major browsers correctly handle cookie strings which specify an empty port (to my knowledge).

Therefore; this should be considered a bug whose severity is proportional to any other bug which affects a widely used fundamental component which mishandles a valid and somewhat common scenario by acting in a cryptic manner.

For instance; in the case of an authorization cookie, the client would simply be incapable of authorizing.
In other scenarios its possible for a webserver to update a cookie to have an empty port at a later time, at which point the cookie becomes non-functional and the data is effectively lost (from a functional perspective).

I think it's very likely that this bug has actually been subtly affecting a wide range of projects and the only reason there have been no bug reports is because affected developers have either assumed it was their own fault, moved to a different cookie handling system, or it happens intermittently and has simply gone unnoticed.

Basically I think it probably affects a lot of people but none have realized that System.Net.Cookie is the broken thing.

So can we change this to a bug and bump up the priority?
Surely you agree with my reasoning here?
